### PR TITLE
Make wesnothd use asio network api instead of SDL_net

### PR DESCRIPTION
**Pull request by loonycyborg**
This removes SDL and SDL_net dependency from wesnothd. This is part of general plan to drop SDL_net, though this PR is only about wesnothd, the multiplayer server.

Reasons for dropping SDL_net are:
- Supporting  IPv6
- Removing dependency on SDL from server components
- Doing networking using proper asynchronous os primitives, which allows avoiding the use of threads and making code less hackish overall.

### CODE REVIEW

**Comment by gfgtdf** [comment]
as notied in irc, there are still some missing codes that need to be impleented before this can be merged specially there are some cases like in 'game::send_data_sides', 'game::send_history' or 'ban_manager::read' which look like they'd break stuff.

**Comment by gfgtdf** [comment]
this has some classes not matching ourt convention of writing everything lin lower letters, like PlayerRecord HandleDoc

**Comment by CelticMinstrel** [comment]
Should the room support be considered a blocker for merging?

**Comment by loonycyborg** [comment]
That depends on whether it will be actually used. AFAIK that feature was only needed for lobby GUI revamp that failed to replace old lobby gui. If someone is working on lobby now would need to ask them whether they absolutely want room support or they would be fine with dropping it.

**Comment by CelticMinstrel** [comment]
I personally think the rooms idea is useful, since it sort of trims the player list to a more manageable length. It doesn't necessarily have to be a blocker for merging though; it could potentially be reimplemented later.

File: src/server/ban.cpp [hunk 1]
```diff
@@ -260,12 +259,12 @@ static lg::log_domain log_server("server");
...
 	void ban_manager::read()
 	{
-		if (filename_.empty() || !filesystem::file_exists(filename_))
+		//if (filename_.empty() || !filesystem::file_exists(filename_))
 			return;
 		LOG_SERVER << "Reading bans from " <<  filename_ << "\n";
 		config cfg;
-		filesystem::scoped_istream ban_file = filesystem::istream_file(filename_);
-		read_gz(cfg, *ban_file);
+		//filesystem::scoped_istream ban_file = filesystem::istream_file(filename_);
```
Comment (aginor) [inline src/server/ban.cpp]: This functionality has been dropped, I think that being able to ban a player is rather important.

File: src/server/player_connection.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/* $Id$ */
+/*
+   Copyright (C) 2012 by Sergey Popov <loonycyborg@gmail.com>
```
Comment (aginor) [inline src/server/player_connection.hpp]: wrong year in comment

File: src/server/server.hpp [hunk 3]
```diff
@@ -17,42 +17,69 @@
...
+#include <boost/asio/signal_set.hpp>
+
+namespace wesnothd
+{
+
 class server
 {
 public:
-	server(int port, const std::string& config_file, size_t min_threads,size_t max_threads);
+	server(int port, bool keep_alive, const std::string& config_file, size_t min_threads,size_t max_threads);
```
Comment (aginor) [inline src/server/server.hpp]: missing space after comma

**Comment by aginor** [comment]
It also looks like this drops support for persisting banned players, I consider that to be a blocker.

File: src/server/game.cpp [hunk 4]
```diff
@@ -71,16 +71,16 @@ namespace
 namespace wesnothd {
 int game::id_num = 1;
 
-void game::missing_user(network::connection socket, const std::string& func) const
+void game::missing_user(socket_ptr /*socket*/, const std::string& func) const
 {
-	WRN_GAME << func << "(): Could not find user (socket:\t" << socket
+	WRN_GAME << func << "(): Could not find user (socket:\t<some C++ pointer>"
```
Comment (aginor) [inline src/server/game.cpp]: This looks like a placeholder comment

File: src/server/server.hpp [hunk 5]
```diff
@@ -115,7 +142,7 @@ class server
 	std::map<std::string,config> proxy_versions_;
 	std::vector<std::string> disallowed_names_;
 	std::string admin_passwd_;
-	std::set<network::connection> admins_;
+	//std::set<network::connection> admins_;
```
Comment (aginor) [inline src/server/server.hpp]: commented out code

File: src/server/server.hpp [hunk 6]
```diff
@@ -87,20 +114,20 @@ class server
 	std::deque<login_log> failed_logins_;
 
 	boost::scoped_ptr<user_handler> user_handler_;
-	std::map<network::connection,std::string> seeds_;
+	std::map<long int,std::string> seeds_;
 
 	/** std::map<network::connection,player>. */
-	wesnothd::player_map players_;
-	std::set<network::connection> ghost_players_;
+	//wesnothd::player_map players_;
```
Comment (aginor) [inline src/server/server.hpp]: commented out code

File: src/util.hpp [hunk 7]
```diff
@@ -496,8 +496,7 @@ inline bool contains(const Container & container, const Value & value)
 
 
 #if 1
```
Comment (aginor) [inline src/util.hpp]: remove #if 1?

File: src/util.hpp [hunk 8]
```diff
@@ -496,8 +496,7 @@ inline bool contains(const Container & container, const Value & value)
...
-# include <SDL_types.h>
-typedef Sint32 fixed_t;
+typedef int32_t fixed_t;
```
Comment (aginor) [inline src/util.hpp]: Why not just use raw type instead of a typedef? There's plenty of precedence of a raw type in the rest of the code with Uint32.

Comment (CelticMinstrel) [inline src/util.hpp]: Why is it called "fixed_t"? To me that implies it's to be used as fixed-point real-number storage; if that's not the case, then @aginor definitely has a good point.

Comment (gfgtdf) [inline src/util.hpp]: Yes it is meant to be used as a fixed point real number. As you can see these is also a 
 typedef float fixed_t; in the #else branch.

I personally don't think this discussion is related to this pr which just replaces Sint32  wuth int32_t to remove a sdl dependency.

File: src/util.hpp (hunk 7 above)
Comment (gfgtdf) [inline src/util.hpp]: Maybe later, but not as part of this pr.

File: src/util.hpp (hunk 8 above)
Comment (CelticMinstrel) [inline src/util.hpp]: Well, personally, I think a fixed point real number is better implemented as a class type (I even implemented it for the formula engine before giving up on that refactor for the time being). But you're right, this isn't really relevant to the PR.

**Comment by gfgtdf** [comment]
Is is ready to be merged?

**Comment by loonycyborg** [comment]
I still need to ensure that forum_user_handler functionality works

**Comment by loonycyborg** [comment]
It's ready to be merged as far as I'm concerned. Perhaps anyone sees something I forgot to do?

**Comment by CelticMinstrel** [comment]
Unless you want to reimplement rooms, but that can probably wait for a later PR.

**Comment by gfgtdf** [comment]
There still seems to be outcommented code in ban_manager::write().

Besides from that i'd be useful to have some comparision in performance of the old vs the new wesnothd like how many games can it handle at the same time on our current hardware.

**Comment by gfgtdf** [comment]
I wonder whether we coudl maybe change the serverside so that the server doesn't need to parse the scenario cfg. So that the cleints sends 2 packages, one with the whole [scenario] and one with the server relevant information, so that the server wouldn't parse (this meant doesn't convert it to a simple_wml structure) the firts packange and just send it to the other players, this might not only improve performance of the wesnothd server it might also fix the translations lost in mp bug since the scenario content is not repackaged.

File: src/server/server.hpp [hunk 9]
```diff
@@ -17,42 +17,70 @@
...
-		send_warning(sock, msg.c_str(), warning_code);
+	boost::asio::io_service io_service_;
+	boost::asio::ip::tcp::acceptor acceptor_;
+	void serve();
+	void accept_connection(const boost::system::error_code& error, socket_ptr socket);
+
+	union {
+		boost::uint32_t connection_num;
+		char buf[4];
+	} handshake_response_;
```
Comment (gfgtdf) [inline src/server/server.hpp]: Hmm this looks like it'd rely on the hardwares endianless here.

Comment (loonycyborg) [inline src/server/server.hpp]: Connection numbers exist only to maintain compatibility with current client. asio_wesnothd doesn't actually use that number for anything else. It keeps track of different client based only on sockets. You see any problems arising if all clients think they're connection number 42, possibly in wrong byte order?

Comment (gfgtdf) [inline src/server/server.hpp]: hmm if its not used than those problems doesnt matter i think.

Comment (loonycyborg) [inline src/server/server.hpp]: It also should directly send/receive gzipped wml sizes but I took care to feed those to htonl()/ntohl()

**Pull request merged**